Create Common Channel Config Bundle to avoid NettyFrameworkImpl's reliance on CHFWBundle

volosied opened the pull request:
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

fixes #19723

Config is now shared between legacy and netty. We will only need to update the new bundle for the config to be available in to transports.

- permissions.perm was deleted due to Tom's request (as it's not needed).
- ConfigurationPolicy.OPTIONAL was changed to REQUIRE as it's necessary configuration (and follows existing patterns)
- chainQuiesceTimeout default was changed in rare scenarios where config isn't picked up 
- ChannelFrameworkConfigImpl cod was mainly copied from the channel fw bundlne
- ChannelFrameworkConfig interface is used to not expose the OSGI methods

Note: Netty only needs the chainQuiesceTimeout, but I implemented all 4 for convenience.

volosied commented:
!build 
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=dd021437-299d-4468-811d-09a32134ce5b) - :x: completed with errors/failures.)
!spawn.fullfat.buckets=com.ibm.ws.transport.http.compression_fat,com.ibm.ws.transport.http.remoteip_fat,com.ibm.ws.transport.http2_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer.servlet.4.0_fat.2,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.3.1_fat.2

_Note: Target locations of links might be accessible only to IBM employees._

helyarp commented:
@volosied Hi, `metatype.properties` has no changes to review. I'll add the **ID review not needed** label. Remove it if you think messages need a review.

volosied commented on dev/com.ibm.ws.channelfw/src/com/ibm/ws/channelfw/internal/ChannelFrameworkImpl.java:
```diff
@@ -164,7 +164,7 @@ public class ChannelFrameworkImpl implements ChannelFramework, FFDCSelfIntrospec
     private long missingConfigWarning = 10000L;
     /** Property for the chain quiescetimeout to default to for various paths in milliseconds */
-    private long chainQuiesceTimeout = 0L;
+    private long chainQuiesceTimeout = 30000L;
```
Since the new policy is `required`, I've set the default to 30s in case it isn't ever picked up.  The other defaults are correct. 
Doc: https://openliberty.io/docs/latest/reference/config/channelfw.html

volosied commented:
#build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=c32a82a1-800d-4f42-9655-232e5dee2df4) - :x: completed with errors/failures.)
#spawn.fullfat.buckets=com.ibm.ws.transport.http.compression_fat,com.ibm.ws.transport.http.remoteip_fat,com.ibm.ws.transport.http2_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer.servlet.4.0_fat.2,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.3.1_fat.2

_Note: Target locations of links might be accessible only to IBM employees._

tjwatson commented on dev/io.openliberty.channel.config.common/resources/permissions.perm:
All our bundles get granted `AllPermission` by default.  I'm not sure why any of our bundles have this OSGi `permissions.perm` file.

tjwatson commented on dev/io.openliberty.channel.config.common/bnd.bnd:
```diff
@@ -0,0 +1,28 @@
+Include-Resource: \
+ OSGI-INF=resources/OSGI-INF, \
+ OSGI-INF/metatype/metatype.xml=resources/OSGI-INF/metatype/metatype.xml, \
+ OSGI-INF/permissions.perm=resources/permissions.perm
```
I think we should stop including this file.  The default is already AllPermission.

tjwatson commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/channel/config/internal/ChannelFrameworkConfig.java:
```diff
@@ -0,0 +1,272 @@
+    /** Custom property for timed delay before warning about missing config in milliseconds */
+    private static long missingConfigWarning = 10000L;
+    /** Property for the chain quiescetimeout to default to for various paths in milliseconds */
+    private static long chainQuiesceTimeout = 30000L;
```
Unless there is no choice we should never manage static state variables (regardless of their type) with the lifecycle of a service component.  Instead the state should live and die with a service component instance of `ChannelFrameworkConfig` and then that service should be injected to where it is needed and non-static methods should be called to access it.

With that in mind, this solution also varies from my original thought which was to have a separate bundle for the metatype AND the service component that holds the state of the config and then that service object is injected to both the netty and the old channel framework components.

volosied replied:
I was modelling off the other existing config bundles, but I like you approach better as we don't need to "re-parse"/ "re-set" the config in every bundle where it's used. 

Though this means we still keep a dependency in Netty, but not on the whole CHFW bundle. Rather depend on the smaller config bundle. 

Let me code up a solution, address permission comment, and get feedback again.  Thanks!

volosied replied:
@tjwatson  Please take another look at the new refactored code. Thanks!

volosied committed:
Create  io.openliberty.channel.config.common bundle for shared channel config

Fix up io.openliberty.channel.config.common

volosied committed:
Update Channel FW to rely on io.openliberty.channel.config.common / ChannelFrameworkConfig

Fix up com.ibm.ws.channelfw

volosied committed:
Update Netty Internal Impl to rely on io.openliberty.channel.config.common / ChannelFrameworkConfig

Fix up io.openliberty.netty.internal.impl

Fix duplicate imports

volosied commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=a5d4ae8a-7130-4af4-8089-19890a7819a5) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

volosied commented:
Small Beta Build: https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_VlMIYLppEfC6CrHd9c_pUg

tjwatson committed:
Handle dynamic configuration changes for channelfw config

tjwatson reviewed (changes requested):
I added my suggested changes to my branch at https://github.com/tjwatson/open-liberty/tree/19723-dependency with commit https://github.com/tjwatson/open-liberty/commit/d9433be28a12fe0327c0d1cf54a833fed574499c

Can you pull in my changes from my fork into your PR?

volosied commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=78f10890-a20e-4fe6-b445-52f0edce3557) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tjwatson reviewed (dismissed):
LGTM

volosied commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/NettyFrameworkImpl.java:
```diff
@@ -190,6 +162,26 @@ protected void setExecutorService(ExecutorService executorService) {
+
+    protected void updatedChannelFWConfig(ChannelFrameworkConfig config) {
+        this.channelConfig = config;
+        // TODO need to do anything so others dynamically react to updates in config?
```
Currently, Netty only needs getDefaultChainQuiesceTimeout, so this TODO is not necessary. However, should I keep it for when we do need perform updates to the code once the config is updated? 

However, if we read from channelConfig, then it should have the latest changes.   

Let me know.

isaacrivriv replied:
IMO I think just updating the config object (like is already done here) should be enough since we use the object on every call for the necessary config.

volosied replied:
I've removed it.

cbridgha reviewed (dismissed):
Looks good

isaacrivriv commented on dev/com.ibm.websphere.appserver.features/visibility/private/com.ibm.websphere.appserver.channelfw-1.0.feature:
```diff
@@ -6,6 +6,7 @@ IBM-Process-Types: server, \
  client
 -bundles= \
  com.ibm.ws.channelfw, \
+ io.openliberty.channel.config.common, \
```
Let's discuss about if the Netty feature eventually should also pull the config bundle in so that we continue removing dependencies on channelfw

volosied replied:
Ah, let me update the Netty feature, too.

volosied replied:
Done -- see [io.openliberty.netty.internal-1.0.feature](https://github.com/OpenLiberty/open-liberty/pull/33262/files#diff-f90f4e1a7f0dae7a36d66da3d5428d3e8921c67ec0d776144e662428945e9216)

tjwatson replied:
Does the build done for this PR not include any tests for netty?  I would have expected massive test failures without this bundle when using netty.

isaacrivriv replied:
As of now for Netty, the private feature is not used for pulling in the bundles given that it would interfere with the GA features which we have this issue as follow up https://github.com/OpenLiberty/open-liberty/issues/25566 once GA. As of the moment, everything that pulls in Netty also pulls in channelfw and so pulls in the config bundle as well and which is why I assume we haven't seen any failures. @volosied did you get a chance to run a Netty build with this?

tjwatson replied:
I am unsure why a private internal feature (e.g. `io.openliberty.netty.internal-1.0`)  could not be used to pull in the netty bundles everywhere you want the netty bundles pulled in.  That should not interfere with GA releases any more than when you include all the netty bundles in the various GA features today.

jhanders34 replied:
I have made the same comment in the past on previous reviews of all of this function.

isaacrivriv replied:
I don't remember the exact error, but the private feature would have to be marked as beta since Netty was delivered as a beta. It would be pulled in by GA features and so give an incompatibility error of beta vs GA. Unless you are saying that we don't need to mark the private feature as beta?

jhanders34 replied:
You would just mark the feature `ga`.  You already have all of those bundles in ga features.  It just makes things much simpler to update things and not need to update everywhere.  The feature is private so isn't a big deal for it to be marked `ga` unless there are auto features that depend on it.

isaacrivriv replied:
I agree it makes life easier, we were just following the beta steps to ensure we are doing everything correctly. I think since we have the issue open we can address this in a different PR to make those changes to use the private feature and make it `ga` if you all agree.

volosied committed:
Address Review Comments:
Update io.openliberty.netty.internal-1.0.feature to use io.openliberty.channel.config.common

Remove TODO as it's not necessary

volosied commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=0f87ea34-5525-4637-850a-b233b972ca80) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

isaacrivriv reviewed (approved):
Changes look good to me, thanks for working on this split!
